// waitkit/dist/index.js
// src/errors.ts
var WaitKitError = class _WaitKitError extends Error {
  constructor(message, statusCode, issues) {
    super(message);
    this.statusCode = statusCode;
    this.issues = issues;
    this.name = "WaitKitError";
  }
  statusCode;
  issues;
  static fromResponse(res) {
    return new _WaitKitError(res.message, res.statusCode, res.issues);
  }
};

// src/client.ts
var HttpClient = class {
  baseUrl;
  constructor(baseUrl) {
    this.baseUrl = baseUrl.replace(/\/+$/, "");
  }
  async request(method, path, options = {}) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (options.query) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const res = await fetch(url.toString(), {
      method,
      headers: {
        "Content-Type": "application/json",
        ...options.headers
      },
      body: options.body ? JSON.stringify(options.body) : void 0
    });
    const json = await res.json();
    if (!res.ok || json.status === "error") {
      const err = json;
      throw WaitKitError.fromResponse(err);
    }
    return json.data;
  }
};

// src/resources/subscribers.ts
var DEFAULT_SOURCE = "landing-page";
function getSourceFromUrl() {
  if (typeof window === "undefined") return null;
  const params = new URLSearchParams(window.location.search);
  return params.get("source") || params.get("s") || null;
}
var SubscribersResource = class {
  client;
  apiKey;
  projectSlug;
  defaultSource;
  constructor(client, apiKey, projectSlug, defaultSource) {
    this.client = client;
    this.apiKey = apiKey;
    this.projectSlug = projectSlug;
    this.defaultSource = defaultSource ?? DEFAULT_SOURCE;
  }
  get headers() {
    return { Authorization: `Bearer ${this.apiKey}` };
  }
  async create(options) {
    const data = await this.client.request(
      "POST",
      `/v1/waitlist/${this.projectSlug}/join`,
      {
        headers: this.headers,
        body: {
          ...options,
          source: options.source ?? getSourceFromUrl() ?? this.defaultSource
        }
      }
    );
    return {
      message: "Thank you for joining the waitlist!",
      email: data.email
    };
  }
  async count(projectSlug) {
    const slug = projectSlug ?? this.projectSlug;
    const data = await this.client.request(
      "GET",
      `/v1/waitlist/${slug}/count`,
      { headers: this.headers }
    );
    return data.count;
  }
};

// src/index.ts
var DEFAULT_BASE_URL = "https://api.waitkit.dev";
var WaitKit = class {
  subscribers;
  client;
  apiKey;
  projectSlug;
  constructor(config) {
    this.apiKey = config.apiKey;
    this.projectSlug = config.projectSlug;
    this.client = new HttpClient(config.baseUrl ?? DEFAULT_BASE_URL);
    this.subscribers = new SubscribersResource(
      this.client,
      this.apiKey,
      this.projectSlug,
      config.source
    );
  }
};
export {
  DEFAULT_BASE_URL,
  WaitKit
};
